### ISSUE TITLE
Memory leaks in Winform controls when using Image Property

### ISSUE THREAD

**Issue by moebassist** [mono/mono#17920]
## Steps to Reproduce

1. Load any GIF into a picture box - larger ones cause the leak faster.
2. Run the code

### Current Behavior

When running on Linux, the application consumes memory until it slows down the whole OS and it's eventually killed. When running on windows, memory use remains constant.

I've attempted to solve the problem in many ways including:

Image Property Setting:

I've split a GIF into separate PNGs, then loaded all the frames into a list. Using a 100ms timer, I iterated through each frame, setting a PictureBox's Image property accordingly.
I confirmed that the PictureBox's Image is being set as a reference to this list - since Disposing the image removes it from the List.

Paint Event Render:

I've attempted to solve the leak by using DrawImage...this yields the same memory leak results.

BackGround Image property:

Using the Background Image....yields the same memory leak results.

Other Controls:

Instead of using a PictureBox, use a label (when using DrawImage in the paint event). This yields the same memory leak results.

### Expected Behavior
Memory use should not climb forever. I don't know where it's disappearing to....
[giftest.zip](https://github.com/mono/mono/files/3891787/giftest.zip)

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

6.4.0.198

**Comment by n432sean** [comment]
I first ran heapshot on giftest, and found no leak with managed memory, suggesting a leak with native code. I then ran the giftest with valgrind and a debug build of libgdiplus 6.0.4, and I found the following set of leaks are repeated.

```
==2342== 21,600 bytes in 1 blocks are definitely lost in loss record 15,550 of 15,682
==2342==    at 0x4C29E63: malloc (vg_replace_malloc.c:309)
==2342==    by 0x10E0158F: GdipAlloc (general.c:155)
==2342==    by 0x10DFCF7C: gdip_bitmap_get_premultiplied_scan0 (bitmap.c:2305)
==2342==    by 0x10DFD0DD: gdip_bitmap_ensure_surface (bitmap.c:2225)
==2342==    by 0x10E15B9D: GdipGetImageGraphicsContext (image.c:320)
==2342==    by 0x14E8788F: ???
==2342==    by 0xCDF1247: ???
==2342==    by 0xCFFFBCF: ???
==2342==    by 0xCFFFBCF: ???
==2342==    by 0x3B: ???
==2342==    by 0xCFFFBCF: ???
==2342==    by 0x6349E4F: ???
==2342== 
==2342== 21,600 bytes in 1 blocks are definitely lost in loss record 15,551 of 15,682
==2342==    at 0x4C29E63: malloc (vg_replace_malloc.c:309)
==2342==    by 0x10E0158F: GdipAlloc (general.c:155)
==2342==    by 0x10DFCF7C: gdip_bitmap_get_premultiplied_scan0 (bitmap.c:2305)
==2342==    by 0x10DFD0DD: gdip_bitmap_ensure_surface (bitmap.c:2225)
==2342==    by 0x10E15E3D: GdipDrawImageRect (image.c:439)
==2342==    by 0x10E16273: GdipDrawImageRect (image.c:406)
==2342==    by 0x10E16352: GdipDrawImageRectI (image.c:376)
==2342==    by 0x14E87F85: ???
==2342==    by 0xCE42F1F: ???
==2342==    by 0x104DBF6F: ???
==2342==    by 0xCE42F47: ???
```

**Comment by n432sean** [comment]
Looks like `gdip_bitmap_ensure_surface` should maybe free `data->scan0` when `premul` is used instead?

**Comment by n432sean** [comment]
The giftest example also leaks a lot of regions
```
==2342== 48 (32 direct, 16 indirect) bytes in 1 blocks are definitely lost in loss record 9,295 of 15,682
==2342==    at 0x4C29E63: malloc (vg_replace_malloc.c:309)
==2342==    by 0x10E0158F: GdipAlloc (general.c:155)
==2342==    by 0x10E2E332: gdip_region_new (region.c:49)
==2342==    by 0x10E2EE78: GdipCreateRegionRect (region.c:636)
==2342==    by 0x10E06AF2: GdipSetClipRect (graphics.c:2000)
==2342==    by 0x10E06B94: GdipSetClipRectI (graphics.c:2011)
==2342==    by 0x14EADCB5: ???
==2342==    by 0x64F2387: ???
==2342==    by 0x1FFEFFF45F: ???
==2342==    by 0xCC05727: ???
==2342==    by 0xE: ???
==2342==    by 0xCFD2767: ???
```

**Comment by n432sean** [comment]
@filipnavara, would you take a look at the leak from gdip_bitmap_get_premultiplied_scan0? I think I have a work around, but I'm not sure of the best way to fix. The other leak from GdipSetClipRect in graphics.c is easy to fix, GdipDeleteRegion (region) just needs to be called before the method exits.

**Comment by Ababuy78** [comment]
Was having similar problem with System.Windows.Forms.
I had an App that was working well on Mono 5.18, when updated to Mono 6.10 the problem started.
When creating (Form.Show()) and distroing forms (Form.Close()) the memory usage just kept increasing, seems like GC wasn't doing it's job.
Tried to destroy controls from the form manually when closing it (it shouldn't be required and as far as I know it is not recommended) and the problem still there.
Tried to maintain the forms on memory and reuse them (Form.Show() and Fomr.Hide()) as needed but every time they are shown again there is an increased memory usage. It seems to be connected to the BackgroundImage property but I am no expert on this.
For me the solution was to revert back to Mono 5.18 because it is was product that needed to be shiped.

**Comment by localuser2** [comment]
I am facing the same Problem: System.Drawing.Bitmap is leaking memory under Linux. Are there any news on the issue?

**Comment by Ababuy78** [comment]
> I am facing the same Problem: System.Drawing.Bitmap is leaking memory under Linux. Are there any news on the issue?

Hi! I built my own forms manager and never unload a form once loaded, only working with the visible property to show / hide forms. My application has a small enough number of forms and the system has enough memory accommodate the total number of forms to work well with this strategy. Hope this helps.

**Comment by localuser2** [comment]
Unfortunately not. Actually it is a Gtk# application. The Bitmap is used as a buffer to "convert" a Windows Forms widget into a Gtk Widget (gtk-dotnet). So, every redraw event leaks memory now.

I am still using Mono 6.8. However, I downgraded libgdiplus to the 5.x version. This seems to work for the moment. Its not a nice solution though.

**Comment by moebassist** [comment]
I’m using winforms. I have a lot of custom drawn controls that override onpaint such as check boxes, buttons, radio options, even a control that converts SVG to c# system.drawing.path because the control rendering is terrible on Debian.  The path control is used to display animations (a different path is drawn each x milliseconds).....this leaked terribly until I replaced ‘invalidate()’ methods in the timer with a full ‘refresh()’. Any use of bitmaps still leak, it seems garbage collection simply fails.

**Comment by moebassist** [comment]
To be clear.....I’ve tried everything to stop leaks with bitmaps....I’m trying to show an animation. I’ve tried:

painting multiple pictures into a single image....leak
Multiple images and toggling visibility....leak
Multiple images and bringtofront()....leak

I think the issue is in the rendering, not the controls since even when I’ve loaded all images into memory and simply change zorder, it leaks.

**Comment by filipnavara** [comment]
I went back through my changes that landed in libgdiplus 6.0.4 to look for any memory leaks. Maybe https://github.com/mono/libgdiplus/pull/635 will fix the problem? Any feedback and testing is welcome.

**Comment by localuser2** [comment]
Okay. I went and did following tests:

- download source of current Debian package of libgdiplus (6.0.4)
- apply patch mono/libgdiplus#635
- build and install modified packages
- no changes, still leaking

- download source of current Debian package of libgdiplus (6.0.4)
- apply all changes from current master branch of libgdiplus to package source (seem to be a lot of additional free this and free that)
- build and install modified packages
- no changes, still leaking

**Comment by n432sean** [comment]
I have a fork with a fix for this leak, but I didn't put up a pull request, because I'm not familiar with this code, and would like for the fix to come from somebody like @filipnavara. The valgrind results that I posted earlier in this thread pinpoint the problems with "premul" not being deleted from gdip_bitmap_ensure_surface and regions not being deleted in GdipSetClipRect.

**Comment by filipnavara** [comment]
@n432sean Thanks for a good pointer. The life time of `premul` from `gdip_bitmap_ensure_surface` is supposed to be aligned with the lifetime of the Cairo surface and the `GpBitmap` object (`System.Drawing.Bitmap` on the manged layer). I fixed part of the issue in https://github.com/mono/libgdiplus/commit/1afe4711f10bd52e14a4c7985eae7414eda4108c but apparently the same thing is still missing in `gdip_bitmap_dispose`. I'll update my PR to fix that.

**Comment by localuser2** [comment]
Will there be a bugfix release for Mono 6.8?

**Comment by filipnavara** [comment]
@localuser2 It is not a bug in Mono, it is a bug in libgdiplus. The release cycles are not necessarily aligned, especially not on Linux.

**Comment by localuser2** [comment]
So, let me rephrase: Will there soon be a bugfix release for libgdiplus which is compatible with Mono 6.8?

**Comment by filipnavara** [comment]
Hopefully yes. Any newer version of libgdiplus should be backwards compatible with Mono 6.8. I'll just have to convince someone to make the release. I've updated the [PR](https://github.com/mono/libgdiplus/pull/635/) based on the observation by @n432sean. If you can test it again it would make me more confident about the fix.

**Comment by localuser2** [comment]
Thanks, I will test again. Did you catch my edit regarding Ubuntu? (last comment)

**Comment by filipnavara** [comment]
I'm not the one making the packages so I cannot give you accurate answer unfortunately. I'll give you an update once the fix is merged.

**Comment by localuser2** [comment]
The fix looks very promising.

Before:
![quickstep_5459](https://user-images.githubusercontent.com/37116544/77938305-d667df00-72b5-11ea-9fd0-cec029f49c64.png)

After:
![quickstep_5613](https://user-images.githubusercontent.com/37116544/77938358-e7b0eb80-72b5-11ea-8c3d-6e610161b24b.png)

Thanks!

**Comment by immarvin** [comment]
Hi, we are suffering from the meamory leak issue in the latest libgdi+ 6.0.4 release. I went through this thread, and found that the issue has been fixed in PR https://github.com/mono/libgdiplus/pull/603 and https://github.com/mono/libgdiplus/pull/635. But seems there is no relese including the fix, so do you have any plan to release a new libgdi+ version with the fix?
